### PR TITLE
gpio implementation discussion with code

### PR DESCRIPTION
**Pull request by seanshpark**
this PR is not for real but for gpio implementation discussion.
please leave comments.
* related issue: Implement GPIO class,  #128

### CODE REVIEW

File: src/js/gpio.js [hunk 1]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
+GPIO.setPin = function(pinNumber, direction, mode, callback) {
+  if (util.isFunction(mode)) {
+    callback = mode;
+    mode = '';
   }
-  else {
-    return err;
+  var dircode = getDirectionCode(direction);
+  if (dircode < 0) {
+    return gpioctl.ERR_INVALIDPARAM;
```
Comment (ILyoan) [inline src/js/gpio.js]: According to GPIO API. we need to call callback function with corresponding `Error` object.
```
var err = new Error("Invalid parameter");
if (util.isFunction(callback)) {
  nextTick(function() {
    callback(err);
  });
} else {
  throw err;
}
```

File: src/js/gpio.js [hunk 2]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
   }
+  var modecode = getModeCode(mode);
+  if (modecode < 0) {
+    return gpioctl.ERR_INVALIDPARAM;
```
Comment (ILyoan) [inline src/js/gpio.js]: Here too. callback with `Error` or throw it in sync mode.

File: src/js/gpio.js [hunk 3]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
+  }
+  return gpioctl.setPin(pinNumber, dircode, modecode, callback);
 };
 
 
-GPIO.write = function(portpin, val, callback) {
-  var err = gpioctl.writepin(portpin, val);
+GPIO.writePin = function(pinNumber, value, callback) {
+  return gpioctl.writepin(pinNumber, value, callback);
```
Comment (ILyoan) [inline src/js/gpio.js]: No return value for this function according to our GPIO API.

File: src/js/gpio.js [hunk 4]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
+};
+
 
-  if (util.isFunction(callback)) {
-    process.nextTick(function() {
-      callback(err);
-    });
+GPIO.readPin = function(pinNumber, callback) {
+  return gpioctl.readPin(pinNumber, callback);
```
Comment (ILyoan) [inline src/js/gpio.js]: No return value for this function according to our GPIO API.

File: src/js/gpio.js [hunk 5]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
+GPIO.setPort = function(portNumber, direction, mode, callback) {
+  if (util.isFunction(mode)) {
+    callback = mode;
+    mode = ' ';
   }
-  else {
-    return err;
+  var dircode = getDirectionCode(direction);
+  if (dircode < 0) {
+    return gpioctl.ERR_INVALIDPARAM;
```
Comment (ILyoan) [inline src/js/gpio.js]: Here too. callback with `Error` or throw it in sync mode.

File: src/js/gpio.js [hunk 6]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
   }
+  return gpioctl.setPort(portNumber, dircode, mode, callback);
 };
 
 
-GPIO.read = function(portpin, callback) {
-  var err = gpioctl.readpin(portpin);
-  var value = err;
+GPIO.writePort = function(portNumber, value, callback) {
+  return gpioctl.writePort(portNumber, value, callback);
```
Comment (ILyoan) [inline src/js/gpio.js]: No return value for this function according to our GPIO API.

File: src/js/gpio.js [hunk 7]
```diff
@@ -47,47 +40,75 @@ GPIO.release = function() {
...
-    process.nextTick(function() {
-      callback(err, value);
-    });
-  }
-  else {
-    return err;
-  }
+
+GPIO.readPort = function(portNumber, callback) {
+  return gpioctl.readPort(portNumber, callback);
```
Comment (ILyoan) [inline src/js/gpio.js]: No return value for this function according to our GPIO API.

File: test/run_pass/test_gpio.js [hunk 8]
```diff
@@ -21,34 +21,28 @@ var assert = require('assert');
...
-});
+result = gpio.setPin(1, "in");
+if (result >= 0)
+  gpioSequence += "A";
 
-gpio.write(0x01, 0xff, function(err) {
-  assert(err>=0, "failed to write");
-  gpioSequence += 'C';
+gpio.setPin(2, "out", function(err) {
+  if (err>=0)
```
Comment (ILyoan) [inline test/run_pass/test_gpio.js]: `err` should be a instance of `Error` ECMA builtin object or `null`.

File: src/iotjs_module_gpioctl.cpp [hunk 9]
```diff
@@ -23,6 +25,57 @@ namespace iotjs {
 
 //-----------------------------------------------------------------------------
 
+GpioCbWrap::GpioCbWrap(JObject& jcallback, gpio_cb_t* cb)
```
Comment (ILyoan) [inline src/iotjs_module_gpioctl.cpp]: Looks like this implementation duplicate with it in "src/iotjs_gpiocbwrap.cpp" ?

File: src/iotjs_gpiocbwrap.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+namespace iotjs {
+
+struct gpio_cb_s {
+  void* data;
+};
+
+typedef struct gpio_cb_s gpio_cb_t;
+
+
+class GpioCbWrap {
```
Comment (ILyoan) [inline src/iotjs_gpiocbwrap.h]: This wrapper looks almost the same with `ReqWrap`.
Could we have general one?

**Comment by seanshpark** [comment]
Thank you for the comments. I'll close this PR when new PR is ready.

**Pull request closed without merging**